src/lib/group.ts

```diff
@@ -0,0 +1,364 @@
+import { Scalar, withMessage } from './core.js';
+import { Field, FieldVar, isField } from './field.js';
+import { Bool, Snarky } from '../snarky.js';
+import { Field as Fp } from '../provable/field-bigint.js';
+import { Pallas } from '../bindings/crypto/elliptic_curve.js';
+
+export { Group };
+
+type FieldLike = FieldVar | Field | number | string | bigint;
+type GroupLike = {
+  x: FieldLike;
+  y: FieldLike;
+};
+
+/**
+ * An element of a Group.
+ */
+class Group {
+  x: Field;
+  y: Field;
+
+  /**
+   * The generator `g` of the Group.
+   */
+  static get generator() {
+    return new Group({ x: Pallas.one.x, y: Pallas.one.y });
+  }
+
+  /**
+   * Coerces anything group-like to a {@link Group}.
+   */
+  constructor({
+    x,
+    y,
+  }: {
+    x: FieldVar | Field | number | string | bigint;
+    y: FieldVar | Field | number | string | bigint;
+  }) {
+    this.x = isField(x) ? x : new Field(x);
+    this.y = isField(y) ? y : new Field(y);
+
+    if (this.#isConstant()) {
+      const { add, mul, square } = Fp;
+
+      let x_bigint = this.x.toBigInt();
+      let y_bigint = this.y.toBigInt();
+
+      let onCurve =
+        add(mul(x_bigint, mul(x_bigint, x_bigint)), Pallas.b) ===
+        square(y_bigint);
+
+      if (!onCurve) {
+        throw Error(
+          `(x: ${x_bigint}, y: ${y_bigint}) is not a valid group element`
+        );
+      }
+    }
+  }
+
+  // helpers
+  static #fromAffine({ x, y }: { x: bigint; y: bigint; infinity: boolean }) {
+    return new Group({ x, y });
```

Comment:
this helper seems a bit dangerous since it's ignoring (and thus, allowing) `infinity` without guaranteeing a canonical `(x, y)` representation of the zero element. I recommend one of two simple options:
- since this seems to be only used once, inline it
- make this method throw when is encounters `infinity: true`